File: Assignments/A10/Assn10.java
import java.io.FileWriter;

class Assn10 {
	public static void main(String[] args) {
		System.out.println("i. Floating point number converter.\n");
		float pi = (float) 3.14159265358979;
		System.out.println(pi + " -> " + convertFloat(pi) + "\n");

		 System.out.println( "\nii. Floating point number enumeration.\n");
		 float fp = (float) 0.0;
		 int i = 0;
		 while (fp < 1.4E-44) {
		 	System.out.println((++i) + ("th number: ") + (fp = nextFloat(fp)));
		 }

		System.out.println("\niii. Floating point number counting\n"); 
		int posFPs = countBetween((float) 0.0, Float.MAX_VALUE);
		int zeroOneFPs = countBetween((float)0.0, (float) 1.0);
		System.out.println("Number of positive floating point numbers: " + posFPs + "\n");
		System.out.println("Number of floating point numbers between 0 and 1: " + zeroOneFPs + "\n");
		System.out.println("Proportion (# of 0~1) / (# of positive): " + ((double) zeroOneFPs / (double) posFPs * 100.0) + "%\n");


		// System.out.println("Args passed: " + args.length);
		if (args.length != 1) {
			System.out.println("\niv. Pass the data file name\n");
		} else { 
			System.out.println("\niv. Floating point number distribution\n");
			FileWriter datafile = null;
			try{
				datafile = new FileWriter(args[1]);
				float lower = (float) 0.0, upper, interval = (float) 100.0;
				for (int j = 0; j < 100; j++) {
					upper = lower + interval;
					datafile.write(countBetween(lower, upper) + "\n");
					lower = upper;
				}
			} catch(Exception e){
				e.printStackTrace();
			} finally{
				try{

				datafile.close();
				} catch (Exception e){
					e.printStackTrace();
				}

			}
			
			System.out.println("The output file is ready. Execute \"Python3 Assn9.py" + args[1] + "\"\n");
		}
	}


	static String convertFloat(float number) {
		String floatVars = null;
		int fBits = Float.floatToRawIntBits(number);   //Converts float to hex-int equiv
        int sign = (fBits & 0x80000000) >> 31;
        int exp = (fBits & 0x7f800000) >> 23;
        int mantissa = (fBits & 0x007fffff);
        floatVars = '(' + Integer.toString(sign) + "," + Integer.toString(exp) + "," + Integer.toString(mantissa) + ')';
		return floatVars;
	}

     static float nextFloat(float number) {
        int fBits = Float.floatToIntBits(number);   //Converts float to hex-int equiv
    	int num = ((fBits & 0x007fffff) + 0x00000001);
    	

        return Float.intBitsToFloat(num);
     }


	static int countBetween(float lower, float upper) {
		int lower_i = Float.floatToIntBits(lower); 
		int upper_i = Float.floatToIntBits(upper);

		return upper_i - lower_i;
	}
}

File: Assignments/A10/Assn10.py
import sys
import struct

def convertFloat (number):
    number_i = struct.unpack('I', struct.pack('f', number))[0]
    sign = (number_i & 0x80000000) >> 31
    exp =  (number_i & 0x7f800000) >> 23
    mantissa = (number_i & 0x007fffff)
    toString = '(' + str(sign) + "," + str(exp) + "," + str(mantissa) + ')'
    return toString 

def nextFloat (number):
    number_i = struct.unpack('I', struct.pack('f', number))[0]
    num = number_i + 1
    return (struct.unpack('f', struct.pack('I', num)))[0]

def countBetween (lower, upper):
    lower = struct.unpack('I', struct.pack('f', lower))[0]
    upper = struct.unpack('I', struct.pack('f', upper))[0]
    return (upper - lower)

print("i. Floating point number converter.")
print(convertFloat(3.14159265358979))

print("\nii. Floating point number enumeration.")
fp = 0.0
i = 0
while fp < 1.4E-44:
    i = i + 1
    fp = nextFloat(fp)
    print (str(i), "th number: ", str(fp))

print("\niii. Floating point number counting")
zeroToMaxFloat = countBetween(0.0, sys.float_info.max)
zeroToOneFloat = countBetween(0.0, 1.0)
print("Number of positive floating point numbers:", str(zeroToMaxFloat))
print("Number of floating point numbers between 0 and 1:", str(zeroToOneFloat))
print("Proportion (# of 0~1) / (# of positive):", str(float(zeroToOneFloat) / float(zeroToMaxFloat) * 100.0), "%")